File: src/main/java/net/volcano/jdacommands/model/command/arguments/implementation/ArgumentParsingData.java
package net.volcano.jdacommands.model.command.arguments.implementation;

import lombok.AllArgsConstructor;
import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
import net.volcano.jdacommands.model.command.Command;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@AllArgsConstructor
public class ArgumentParsingData {
	
	private static final Pattern argumentPattern = Pattern.compile("\"[^\"]+\"|'[^']+'|\\S+", Pattern.DOTALL | Pattern.CASE_INSENSITIVE | Pattern.MULTILINE);
	
	public int currentArg = 0;
	
	public RawArgument[] rawArguments;
	
	public String rawPrefix;
	
	public String rawPath;
	
	public String rawContent;
	
	public MessageReceivedEvent event;
	
	public Command command;
	
	public ArgumentParsingData(MessageReceivedEvent event, String rawPrefix, String rawPath, String rawContent) {
		this.rawPrefix = rawPrefix;
		this.rawPath = rawPath;
		this.rawContent = rawContent;
		this.event = event;
		
		List<RawArgument> argumentList = new ArrayList<>();
		Matcher matcher = argumentPattern.matcher(rawContent.trim());
		
		while (matcher.find()) {
			var rawArgumentBuilder = RawArgument.builder();
			
			String token = matcher.group();
			if (token.matches("^(['\"].*['\"])$") &&
					token.length() > 2) {
				rawArgumentBuilder.inQuotations(true);
				rawArgumentBuilder.value(token.substring(1, token.length() - 1));
			} else {
				rawArgumentBuilder.inQuotations(false);
				rawArgumentBuilder.value(token);
			}
			rawArgumentBuilder.startIndex(matcher.start());
			argumentList.add(rawArgumentBuilder.build());
		}
		rawArguments = argumentList.toArray(new RawArgument[0]);
		
	}
	
	public boolean nextArgument() {
		if (hasNext()) {
			currentArg++;
			return true;
		}
		return false;
	}
	
	public boolean hasNext() {
		return currentArg + 1 < size();
	}
	
	public int size() {
		return rawArguments.length;
	}
	
	public String getArg() {
		return rawArguments.length > currentArg ? rawArguments[currentArg].value : null;
	}
	
	@Override
	public ArgumentParsingData clone() {
		return new ArgumentParsingData(
				currentArg,
				rawArguments,
				rawPrefix,
				rawPath,
				rawContent,
				event,
				command
		);
	}
	
}

File: src/main/java/net/volcano/jdacommands/model/command/arguments/implementation/ArgumentParsingData.kt
package net.volcano.jdacommands.model.command.arguments.implementation

import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import net.volcano.jdacommands.model.command.Command
import net.volcano.jdautils.utils.TOKEN_REGEX

val ARGUMENT_REGEX =
	Regex(TOKEN_REGEX, setOf(RegexOption.DOT_MATCHES_ALL, RegexOption.MULTILINE, RegexOption.IGNORE_CASE))

class ArgumentParsingData(
	var event: MessageReceivedEvent,
	var rawPrefix: String,
	var rawPath: String,
	var rawContent: String
) {

	var currentArg = 0

	var rawArguments: Array<RawArgument>

	lateinit var command: Command

	fun nextArgument(): Boolean {
		if (hasNext()) {
			currentArg++
			return true
		}
		return false
	}

	operator fun hasNext(): Boolean {
		return currentArg + 1 < size
	}

	val size: Int
		get() = rawArguments.size

	val arg: String?
		get() = if (rawArguments.size > currentArg) rawArguments[currentArg].value else null

	fun clone(): ArgumentParsingData {
		val data = ArgumentParsingData(
			event,
			rawPrefix,
			rawPath,
			rawContent
		)
		data.currentArg = currentArg
		data.command = command
		return data
	}

	init {
		val argumentList: MutableList<RawArgument> = ArrayList()
		for (result in ARGUMENT_REGEX.findAll(rawContent.trim { it <= ' ' })) {
			val token = result.value
			if (Regex("^(['\"].*['\"])$").matches(token) &&
				token.length > 2
			) {
				argumentList.add(RawArgument(token.substring(1, token.length - 1), result.range.first, true))
			} else {
				argumentList.add(RawArgument(token, result.range.first, false))
			}
		}
		rawArguments = argumentList.toTypedArray()
	}
}